core/src/main/java/net/opentsdb/query/execution/StorageQueryExecutor.java

```diff
@@ -0,0 +1,167 @@
+  private TimeSeriesDataStore data_store;
+  
+  /**
+   * Default ctor.
+   * @param node A non-null node to pull the ID and config from.
+   * @throws IllegalArgumentException if a config param was invalid.
+   */
+  public StorageQueryExecutor(final ExecutionGraphNode node) {
+    super(node);
+    if (node.getDefaultConfig() == null) {
+      throw new IllegalArgumentException("Default config cannot be null.");
+    }
+    if (node.graph() == null) {
+      throw new IllegalStateException("Execution graph cannot be null.");
+    }
+    if (Strings.isNullOrEmpty(((Config) node.getDefaultConfig()).getStorageId())) {
+      throw new IllegalArgumentException("Default storage ID must be non-empty.");
+    }
+    data_store = (TimeSeriesDataStore) node.graph().tsdb().getRegistry()
+        .getPlugin(TimeSeriesDataStore.class, 
+            ((Config) node.getDefaultConfig()).getStorageId());
+    if (data_store == null) {
+      throw new IllegalStateException("No data store found for ID " 
+          + ((Config) node.getDefaultConfig()).getStorageId());
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public QueryExecution<T> executeQuery(final QueryContext context,
+                                        final TimeSeriesQuery query, 
+                                        final Span upstream_span) {
+    
+    class Cleanup implements Callback<T, T> {
+      private final QueryExecution<T> execution;
+      public Cleanup(final QueryExecution<T> execution) {
+        this.execution = execution;
+      }
+      @Override
+      public T call(final T arg) throws Exception {
+        outstanding_executions.remove(execution);
+        return arg;
+      }
+    }
+    
+    final QueryExecution<T> execution = 
+        (QueryExecution<T>) data_store.runTimeSeriesQuery(context, query, upstream_span);
+    outstanding_executions.add(execution);
+    execution.deferred().addBoth(new Cleanup(execution));
+    return execution;
+  }
+
+  /** The configuration class for this executor. */
+  @JsonInclude(Include.NON_NULL)
+  @JsonDeserialize(builder = Config.Builder.class)
+  public static class Config extends QueryExecutorConfig {
+    private String storage_id;
+    
+    protected Config(Builder builder) {
+      super(builder);
+      storage_id = builder.storageId; 
+    }
+    
+    public String getStorageId() {
+      return storage_id;
+    }
+    
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      final Config config = (Config) o;
+      return Objects.equal(executor_id, config.executor_id)
+          && Objects.equal(executor_type, config.executor_type)
+          && Objects.equal(storage_id, config.storage_id);
+    }
+    
+    @Override
+    public int hashCode() {
+      return buildHashCode().asInt();
+    }
+
+    @Override
+    public HashCode buildHashCode() {
```

Comment:
Please document why these specific fields were chosen for hashing, equals etc..

Reply:
Doc'd in the interface. I think.